### DISCUSSION TITLE
[1.1.446] Effective region mask for ControlNet/IPAdapter

### DISCUSSION THREAD

**Discussion by huchenlei** [Mikubill/sd-webui-controlnet#2831] [General]
Previously mask upload was only used as an alternative way for user to specify more precise mask. It was only supported for inpaint and ipadapter CLIP mask. See answer to https://github.com/Mikubill/sd-webui-controlnet/discussions/2793.

However, this feature seems to be under-used. We observe that SD Forge uses the mask upload UI to specify effective region. (Note: I am not calling it attn mask as I feel it misleading to the attn mask in transformer argument). In 1.1.446, we implemented the UI to allow people use the UI the same way forge does.

Useage example:
Previously there is no way to prevent depth controlnet to not affect part of the image. If you use pure black to fill the depth map, the result will not be satisfying.
![30 04 2024_22 49 14_REC](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/b950ed9d-f2cd-4c34-a698-5f2e93e3365e)
![00013-81378068](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/f601f953-9362-4d44-82e8-1bde4b1ad9d9)

However with effective region mask, now you can limit the ControlNet effect to certain part of image. Here we are only allowing depth controlnet to control left part of image.
![30 04 2024_22 46 51_REC](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/7453469c-5959-4b55-ab16-0a22ad425c28)
![grid-0002](https://github.com/Mikubill/sd-webui-controlnet/assets/20929282/e6494cd6-7a1a-4101-8e82-1053332c697b)

There will be a more user friendly region planner tool later to help layout control region for different controlnet unit. https://github.com/Mikubill/sd-webui-controlnet/pull/2823. Please stay tuned.

**Comment by a3nima** [comment]
This is really cool. Any tips on how to use it with IP Adapter Face ID? Seems to work kind of, but not as well as with only 1 controlnet.

**Comment by huangjianyi0701** [comment]
Hello, I got an error during the use of mask,
 the error is as follows:
 
![image](https://github.com/user-attachments/assets/4184d85c-893e-49a3-b99b-219329e35c64)
![image](https://github.com/user-attachments/assets/4c8e3b04-a176-456a-b799-e1f847f21e39)

 The environment is as follows:
sd:1.8
controlnet:1.1.455
torch:2.4.1
cuda:12.2

**Reply by huangjianyi0701** [reply]
![image](https://github.com/user-attachments/assets/e1a9b1d7-1c0d-499b-8361-e738dc345cd3)

**Reply by huangjianyi0701** [reply]
I found that your method works when the diagram is born, when the mask is redrawn, the mask uploaded by ipadapter will not take effect and report an error, he will enter this function three times, and the inconsistency of the dimensions of the out and mask at the third output will result in an error, what is the problem please?

**Reply by kasalelia** [reply]
Try this, set output Width=Height, and >800, when use Effective Region Mask
